### ISSUE TITLE
IPLD requirements for js-ipfs

### ISSUE THREAD

**Issue by mikeal** [ipfs/js-ipfs#1893]
The IPLD team is trying to get a handle on our roadmap and plan for the next few quarters. js-ipfs is the primary consumer of JS IPLD interfaces and we’d like to get a good list of things js-ipfs needs over the next year or so.

**Comment by daviddias** [comment]
IPLD packs a ton of promises for the future of building distributed computing. However, rather than giving a long list of all future hopes and dreams, I'll make one request that reflects one current use case of IPLD over js-ipfs other than IPFS using it for UnixFS.

Can js-ipld reproducibility and correctly resolve data in formats used by blockchains such as the Ethereum Blockchain? This was possible before and it created a ton of momentum and value, things to note are:

- We made an example with it https://github.com/ipfs/js-ipfs/tree/master/examples/explore-ethereum-blockchain
- We made a huge demo -- https://www.youtube.com/watch?v=HxueJbeMVG4 --. It enphasises libp2p but the reality is that it was only possible thanks to IPLD
- Took notes on next steps https://github.com/ipfs/notes/issues/173
- Ran a call for a while https://github.com/ipfs/team-mgmt/issues/484
- Got a bridge from Ethereum to IPFS working using the Parity client -- https://github.com/ipfs/js-ipfs/issues/763
- A new project, Mustekala, got funded and developed https://medium.com/metamask/metamask-labs-presents-mustekala-the-light-client-that-seeds-data-full-nodes-vs-light-clients-3bc785307ef5. 

It would be really great to have this stable and something that we can point people at as a great example of interoperability

**Comment by daviddias** [comment]
Ah, another great example of interoperability between networks and formats is https://github.com/ipfs-shipyard/igis, right now it lacks the maintenance necessary.

**Comment by daviddias** [comment]
One more ask (😇) support for Torrents https://github.com/ipfs/js-ipfs/issues/779

**Comment by daviddias** [comment]
> you added repos for those implementations to https://github.com/ipld/ipld#implementations but apparently those are just empty repos 😱

Ups, sorry! I did a lot of experimentation locally and found some obstacles along the way, then my attention was routed to other things (it was mid 2017 :)). I left a list of things to be done at https://github.com/ipfs/js-ipfs/issues/779 

> Writing JS codecs for torrents would be pretty easy, I’ve spent time a fair amount of time in those torrent libraries already.

True, but depending on how it is integrated, it gets tricky. A torrent file is referenced by the hash (called infohash in torrent world) of the file that the torrent is sharing, but not the torrent file itself. torrent files keep mutating as new trackers get added.

Adding support to resolve torrent file chunks to IPLD should be trivial, I agree, however, in the ideal world one should be able to do something like `ipfs torrent cat /infohash` and it would pipe out the contents of the torrent by connecting to the torrent network and piecing the file together with bits existing there and other bits already existing in the IPFS network, if available.

> I’m not going to consider this high priority (compared to unixfsv2) but this is something small and fun and I’ll probably get to it in Q4 cause I love you @daviddias 😘

I didn't mean to derail the plan for Q4, but this does make me super happy, thaaank you ❤!

> ETH

Understood. It would be good to:

- Ensure that the example is running perfectly
- Potentially re record a video demo that can be shared
- Reach out to @kumavis and @hermanjunge to learn what is the latest state of Mustekala. I recall that for the Blockchain usecase, Bitswap was not ideal and now with Graphsync, it can potentially make the use case much more powerful.

**Comment by mikeal** [comment]
> True, but depending on how it is integrated, it gets tricky. A torrent file is referenced by the hash (called infohash in torrent world) of the file that the torrent is sharing, but not the torrent file itself. torrent files keep mutating as new trackers get added.

Ya, I had read the prior threads and I’m already taking for granted the `torrent-file` + `torrent` approach at the IPLD layer.

> Adding support to resolve torrent file chunks to IPLD should be trivial, I agree, however, in the ideal world one should be able to do something like ipfs torrent cat /infohash and it would pipe out the contents of the torrent by connecting to the torrent network and piecing the file together with bits existing there and other bits already existing in the IPFS network, if available.

As a first step, we should just get `ipfs cat CID-of-torrent-file` working.

Rather than infohash, why don’t you just use a magnet URL? We could even add a parameter to the magnet URL for the CID of the torrent-file, then it would contain everything we need for getting the file out of both networks. Maybe this sounds like cheating, but from a user perspective it’s already common to prefer magnet urls over infohashes, and you’ve already stubbed out a separate command in your example for getting torrents so the parameter doesn’t have to conform to multiaddr.

The networking side of this is outside the scope of IPLD but I can try to advise where I can since I have some experience with these JS libraries and with bittorrent in general.

**Comment by mikeal** [comment]
I’ve thought a lot about the “patch” API and I’m still not sure it’s possible to do something atomic **and** performant.

A patch API, or any mutation API, needs to be efficiently batched in order to be performant when manipulating the same parts of the DAG concurrently. This is actually something I plan on writing soon on top of unixfsv2 (for mfs), taking all concurrent mutation operations and batching them into one large graph mutation that doesn’t unnecessarily create/read any part of the tree more than once.

I can produce this as a standalone library for efficient graph manipulation, but the API will have to sit on top of a larger mutex used for batching concurrent operations and I’m not sure how well that would fit into `js-ipfs`’s current architecture. It would have an API along the lines of:

```javascript
const m = mutex(rootNode, { /* storage functions */ getBlock, setBlocks })
await m.set(‘/path/to/node’, newBlock)
```

You *could* use this atomically by just creating a new mutex for every operation, but it won’t be nearly as performant as keeping a single instance around and hammering it with concurrent mutations like I’ll do with the mfs code.

In short, you’ll have something for this by the end of the quarter, but it may take larger changes in `js-ipfs` in order to make it actually perform well.

**Comment by SgtPooki** [comment]
js-ipfs is being deprecated in favor of [Helia](https://github.com/ipfs/helia).  You can [learn more about this deprecation](https://github.com/ipfs/js-ipfs/issues/4336) and [read the migration guide](https://github.com/ipfs/helia/wiki/%F0%9F%9A%9B-Migrating-from-js-IPFS).  

Please feel to reopen with any comments by 2023-06-02.  We will do a final pass on reopened issues afterwards (see https://github.com/ipfs/js-ipfs/issues/4336).
